src/WinRT.Runtime/ComWrappersSupport.cs

```diff
@@ -168,72 +176,75 @@ internal static ObjectReference<T> GetObjectReferenceForInterface<T>(IntPtr exte
             bool hasCustomIMarshalInterface = false;
+            bool hasWinrtExposedClassAttribute = false;
 
-            if (type.IsDelegate())
+#if NET
+            // Check whether the type itself has the WinRTTypeExposed attribute and if so
+            // use the new source generator approach.
+            var winrtExposedClassAttribute = type.GetCustomAttribute<WinRTExposedTypeAttribute>(false);
+
+            // Handle scenario where it can be an authored type
+            // which means the attribute lives on the authoring metadata type.
+            if (winrtExposedClassAttribute == null)
             {
-                // Delegates have no interfaces that they implement, so adding default WinRT entries.
-                var helperType = type.FindHelperType();
-                if (helperType is object)
+                var authoringMetadaType = type.GetRuntimeClassCCWType();
+                if (authoringMetadaType != null)
                 {
-                    entries.Add(new ComInterfaceEntry
-                    {
-                        IID = GuidGenerator.GetIID(type),
-                        Vtable = helperType.GetAbiToProjectionVftblPtr()
-                    });
+                    winrtExposedClassAttribute = authoringMetadaType.GetCustomAttribute<WinRTExposedTypeAttribute>(false);
                 }
+            }
+
+            if (winrtExposedClassAttribute != null)
+            {
+                hasWinrtExposedClassAttribute = true;
+                entries.AddRange(winrtExposedClassAttribute.GetExposedInterfaces());
 
-                if (type.ShouldProvideIReference())
+                if (type.IsClass)
                 {
-                    entries.Add(IPropertyValueEntry);
-                    entries.Add(ProvideIReference(type));
+                    hasCustomIMarshalInterface = entries.Any(entry => entry.IID == ABI.WinRT.Interop.IMarshal.IID);
                 }
             }
-            else
+            else if (type == typeof(global::System.EventHandler))
+            {
+                hasWinrtExposedClassAttribute = true;
+                entries.AddRange(ABI.System.EventHandler.GetExposedInterfaces());
+            }
+            else if (!type.IsEnum && GetComInterfaceEntriesForTypeFromLookupTable(type) is var lookupTableEntries && lookupTableEntries != null)
             {
-                var objType = type.GetRuntimeClassCCWType() ?? type;
-                // Check whether the type itself has the attribute to make sure it is using the new source generator
-                // and just doesn't have an updated dependent projection which is.
-                if (type.GetCustomAttribute(typeof(WinRTExposedTypeAttribute), false) is IWinRTExposedTypeDetails winrtExposedClassAttribute)
+                hasWinrtExposedClassAttribute = true;
+                entries.AddRange(lookupTableEntries);
+            }
+            else if (RuntimeFeature.IsDynamicCodeCompiled)

```

Comment:
Is it possible for this path to be-AOT compatible somehow? This change can cause some IIDs to be missing entirely, causing `CreateCCWForObjectForABI` to fail. One affected API is `StorageFile.GetFileFromPathAsync`.

Here's a screenshot of `AsyncOperationCompletedHandler<StorageFile>` in CoreCLR (left) versus NativeAOT (right) (provided by @hez2010):
![Screenshot of two Windows Terminal panes. The GUIDs of `AsyncOperationCompletedHandler<StorageFile>` are printed in both CoreCLR and NativeAOT environments](https://github.com/microsoft/CsWinRT/assets/29563098/4553b368-84a3-469a-bf7d-a9a733c3974e)

Reply:
This code path is meant to be the fallback.  It looks like the first guid is the one that is missing with the other extra ones being duplicates.  For this scenario, the WinRT AOT source generator should have generated the necessary vtable entries as part of the lookup table.  Do you see any warnings for `WinRTAotSourceGenerator` failing to run?  If not, are you able to check if there is a generated file called WinRTGlobalVtableLookup.g.cs and inside of it, do you see a reference to the mentioned type?

Reply:
Those missing IIDs are delegates, such as `AsyncOperationCompletedHandler<T>`.

Reply:
I checked the `WinRTGlobalVtableLookup.g.cs`, it did exist but I didn't find the code generated for `Windows.Foundation.AsyncOperationCompletedHandler<Windows.Storage.StorageFile>`.

Reply:
I added a test case to test this scenario and I discovered a different issue, but not this specific one you are running into.  I think I know why you are seeing this.  Today, you typically add a CsWinRT package reference to generate a projection for a set of WinMD files or for authoring scenarios.  With these changes, you also need to add a CsWinRT package reference to make your library or app which consumes CsWinRT projected APIs `AOT compatible`.  This is because we have a new source generator that we are using to generate some code in order to be AOT compatible specifically in the generic type scenarios.  In the future we may look for over ways to have the source generator run without needing a CsWinRT package reference such as when using the net windows10 TFM.

Can you confirm whether your app or library where this `AsyncOperationCompletedHandler` is being added has a CsWinRT package reference.  If not, can you add one and see if that helps.  I assume you are testing this by manually building my changes, so you might not have a CsWinRT package built.  If so, you might be able to directly add a reference to the source generator as we do in our functional tests [here](https://github.com/microsoft/CsWinRT/blob/2a7b7548b237597e32de9dcfcac0927bb6fc6ff3/src/Tests/FunctionalTests/Async/Async.csproj#L12).  Let me know if that addresses it.

Reply:
I didn't see your last comment when I responded.  Was the `WinRTGlobalVtableLookup.g.cs` file you found for the projection or for your library / app.

Reply:
It's for the projection.

Reply:
Okay now I tested with the source generator, but seems that it's still failing as `OpenStreamForReadAsync` is not a method from WinRT types, so nothing was generated for ``Windows.Foundation.AsyncOperationCompletedHandler`1[Windows.Storage.Streams.IRandomAccessStream]`` as this type is used by `OpenStreamForReadAsync` which is not available as source code.

```cs
var file = await StorageFile.GetFileFromPathAsync(...);
using var stream = await file.OpenStreamForReadAsync();
using var sw = new StreamReader(stream);
Console.WriteLine(sw.ReadToEnd());
```

```
Unhandled Exception: System.InvalidCastException: Specified cast is not valid.
   at WinRT.ComWrappersSupport.CreateCCWForObjectForABI(Object, Guid) + 0xb0
   at WinRT.MarshalDelegate.CreateMarshaler2(Object, Guid, Boolean) + 0x76
   at ABI.Windows.Foundation.AsyncOperationCompletedHandler`1.CreateMarshaler2(AsyncOperationCompletedHandler`1 managedDelegate) + 0x4d
   at WinRT.GenericTypeInstantiations.Windows_Foundation_IAsyncOperation_1_Windows_Storage_Streams_IRandomAccessStream.put_Completed(IObjectReference, AsyncOperationCompletedHandler`1) + 0x63
   at System.WindowsRuntimeSystemExtensions.AsTask[TResult](IAsyncOperation`1, CancellationToken) + 0x130
   at System.IO.WindowsRuntimeStorageExtensions.<OpenStreamForReadAsyncCore>d__1.MoveNext() + 0x49
```

The generated code by the source generator is:

```cs
namespace WinRT.GenericHelpers
{

// Windows.Foundation.AsyncOperationCompletedHandler<global::Windows.Storage.StorageFile>
internal static class AsyncOperationCompletedHandler_Windows_Storage_StorageFile
{
    private static bool _initialized = Init();
    internal static bool Initialized => _initialized;

    private static unsafe bool Init()
    {
        return global::ABI.Windows.Foundation.AsyncOperationCompletedHandlerMethods<Windows.Storage.StorageFile, IntPtr>.InitCcw(
           &Do_Abi_Invoke
        );
    }

    [UnmanagedCallersOnly(CallConvs = new[] { typeof(CallConvStdcall) })]
    private static unsafe int Do_Abi_Invoke(IntPtr thisPtr, IntPtr asyncInfo, global::Windows.Foundation.AsyncStatus asyncStatus)
    {
        try
        {
            global::ABI.Windows.Foundation.AsyncOperationCompletedHandlerMethods<Windows.Storage.StorageFile, IntPtr>.Abi_Invoke(thisPtr, MarshalInterface<global::Windows.Foundation.IAsyncOperation<Windows.Storage.StorageFile>>.FromAbi(asyncInfo), asyncStatus);
        }
        catch (global::System.Exception __exception__)
        {
            global::WinRT.ExceptionHelpers.SetErrorInfo(__exception__);
            return global::WinRT.ExceptionHelpers.GetHRForException(__exception__);
        }
        return 0;
    }
}
}

namespace WinRT.GenericHelpers
{

    internal static class GlobalVtableLookup
    {

        [System.Runtime.CompilerServices.ModuleInitializer]
        internal static void InitializeGlobalVtableLookup()
        {
            ComWrappersSupport.RegisterTypeComInterfaceEntriesLookup(LookupVtableEntries);
        }

        private static ComWrappers.ComInterfaceEntry[] LookupVtableEntries(Type type)
        {
            if (type == typeof(global::Windows.Foundation.AsyncOperationCompletedHandler<global::Windows.Storage.StorageFile>))
            {
                        _ = global::WinRT.GenericHelpers.AsyncOperationCompletedHandler_Windows_Storage_StorageFile.Initialized;

        var delegateInterface = new global::System.Runtime.InteropServices.ComWrappers.ComInterfaceEntry
        {
            IID = global::WinRT.GuidGenerator.GetIID(typeof(global::Windows.Foundation.AsyncOperationCompletedHandler<global::Windows.Storage.StorageFile>).GetHelperType()),
            Vtable = global::ABI.Windows.Foundation.AsyncOperationCompletedHandler<global::Windows.Storage.StorageFile>.AbiToProjectionVftablePtr
        };

        return global::WinRT.DelegateTypeDetails<Windows.Foundation.AsyncOperationCompletedHandler<global::Windows.Storage.StorageFile>>.GetExposedInterfaces(delegateInterface);

            }
            return default;
        }
    }
}

```

Reply:
Also note that the source generator doesn't work for code like:

```cs
var file = await StorageFile.GetFileFromPathAsync(...);
using var stream = await file.OpenAsync(FileAccessMode.Read).AsTask().ConfigureAwait(continueOnCapturedContext: false);
using var sw = new StreamReader(stream.AsStreamForRead());
Console.WriteLine(sw.ReadToEnd());
```

You have to write it like the below to make the source generator work:

```cs
var file = await StorageFile.GetFileFromPathAsync(...);
using var stream = await file.OpenAsync(FileAccessMode.Read);
using var sw = new StreamReader(stream.AsStreamForRead());
Console.WriteLine(sw.ReadToEnd());
```

Reply:
Thanks for confirming.  `OpenStreamForReadAsync` should be coming from [here](https://github.com/microsoft/CsWinRT/blob/71169932920500310921b4c3fadba1e5a99c6138/src/cswinrt/strings/additions/Windows.Storage/WindowsRuntimeStorageExtensions.cs#L28).  I will need to enlighten the source generator to handle this scenario.  I will also look into the other scenario you mentioned which it doesn't handle.

Reply:
I think you can generate code for async operations at the invocation instead of the await expression:

i.e. adding the below code under `if (invocationSymbol is IMethodSymbol methodSymbol && GeneratorHelper.IsWinRTType(methodSymbol.ContainingSymbol))` in `GetVtableAttributesToAddOnLookupTable`:
```cs

                    if (GeneratorHelper.IsWinRTType(methodSymbol.ReturnType))
                    {
                        // if (!System.Diagnostics.Debugger.IsAttached) System.Diagnostics.Debugger.Launch();
                        var completedProperty = methodSymbol.ReturnType.GetMembers("Completed").FirstOrDefault() as IPropertySymbol;
                        if (completedProperty?.Type.MetadataName.Contains("`") ?? false)
                        {
                            vtableAttributes.Add(GetVtableAttributeToAdd(completedProperty.Type, GeneratorHelper.IsWinRTType, context.SemanticModel.Compilation.Assembly, false));
                        }
                    }
```

Reply:
I created a PR for you to resolve the above issue: https://github.com/microsoft/CsWinRT/pull/1383

Reply:
Partially unrelated, but I'm very confused by this line in the generated code above:
```csharp
IID = global::WinRT.GuidGenerator.GetIID(typeof(global::Windows.Foundation.AsyncOperationCompletedHandler<global::Windows.Storage.StorageFile>).GetHelperType())
```
Given this is all generated code, can't the generator precompute that IID into some static field as usual (like all new AOT-friendly projections are also doing in this PR) and just use that here, rather than using `GuidGenerator`?

Reply:
I left this for now as I have plans to try to remove the delegates from the lookup table if possible and move it into the projection itself similar to the rcw instantiations.